lib/ts/recipe/session/recipeImplementation.ts

```diff
@@ -267,41 +274,36 @@ export default class RecipeImplementation implements RecipeInterface {
             }
             ProcessState.getInstance().addState(PROCESS_STATE.CALLING_SERVICE_IN_GET_HANDSHAKE_INFO);
             let response = await this.querier.sendPostRequest(new NormalisedURLPath("/recipe/handshake"), {});
-            let signingKeyLastUpdated = Date.now();
-            if (this.handshakeInfo !== undefined) {
-                if (
-                    response.jwtSigningPublicKeyExpiryTime === this.handshakeInfo.jwtSigningPublicKeyExpiryTime &&
-                    response.jwtSigningPublicKey === this.handshakeInfo.jwtSigningPublicKey
-                ) {
-                    signingKeyLastUpdated = this.handshakeInfo.signingKeyLastUpdated;
-                }
-            }
+
             this.handshakeInfo = {
-                signingKeyLastUpdated,
-                jwtSigningPublicKey: response.jwtSigningPublicKey,
+                jwtSigningPublicKeyList: response.jwtSigningPublicKeyList,
                 antiCsrf,
                 accessTokenBlacklistingEnabled: response.accessTokenBlacklistingEnabled,
-                jwtSigningPublicKeyExpiryTime: response.jwtSigningPublicKeyExpiryTime,
                 accessTokenValidity: response.accessTokenValidity,
                 refreshTokenValidity: response.refreshTokenValidity,
             };
+
+            this.updateJwtSigningPublicKeyInfo(
+                response.jwtSigningPublicKeyList,
+                response.jwtSigningPublicKey,
+                response.jwtSigningPublicKeyExpiryTime
+            );
+
             if (this.isInServerlessEnv) {
                 storeIntoTempFolderForServerlessCache(SERVERLESS_CACHE_HANDSHAKE_INFO_FILE_PATH, this.handshakeInfo);
             }
         }
         return this.handshakeInfo;
     };
 
-    updateJwtSigningPublicKeyInfo = (newKey: string, newExpiry: number) => {
+    updateJwtSigningPublicKeyInfo = (keyList: KeyInfo[] | undefined, publicKey: string, expiryTime: number) => {
+        if (keyList === undefined) {
```

Comment:
or the length of it is 0? Cause when will it ever be `undefined`?

Reply:
Before 2.9, it's always undefined. With 2.9, it always contains at least one key.

Reply:
Alright. Please add a comment for this.

Reply:
@porcellus perhaps you missed this comment.

Reply:
I added the comment right above the function, I think GitHub is just displaying it weirdly.

Reply:
Ah fair enough.. my bad. Will resolve this comment